### PR TITLE
Migration Tutorial: From Public to Cloud

### PR DESCRIPTION
**Pull request by akosma**
This PR provides a solution for issue #39.

## Checklist

- [x] Categorize the PR by setting a good title and adding one of the labels:
      `how-to`, `tutorial`, `reference`, `explanation`, `cicd`, `dependency`
      as they show up in the changelog
- [x] Link this PR to related issues if applicable.

### CODE REVIEW

**Comment by akosma** [comment]
I've started by copying the text provided by @megian in other issues, will rewrite and adapt next.

**Comment by akosma** [comment]
@tobru @megian @mhutter please review, thanks!

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (tobru) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:9]: I don't think that the Timezone is the most important difference to especially highlight here. I'd rather improve the `differences-to-public.adoc` regarding that topic.

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (tobru) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:21]: 4.9 would be 1.22 (https://docs.openshift.com/container-platform/4.9/release_notes/ocp-4-9-release-notes.html#ocp-4-9-about-this-release)

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (tobru) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:27]:
```suggestion
. Lowering your TTLs (Time To Live) 24 hours ahead of time will reduce the amount of time that your change takes to propagate. You can limit downtime by lowering that value, so the caching servers will check back more frequently. Typically, 60 seconds (1 minute) is a good idea.
```

Note: There is no default of TTL.

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (tobru) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:29]: I'm not sure how much value this example brings to this stage, I'd just remove it.

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (tobru) [inline docs/modules/ROOT/pages/tutorials/migration.adoc]:
```suggestion
. There are different `CNAME` resource records to point to depending on which {zone} you migrated the application to; please refer to xref:zones.adoc[the {zone}s] reference document, where the corresponding `CNAME` entry is specified.
```

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (tobru) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:116]: As we'll be enabling RWX storage in the next two weeks, I'd rather not mention that here as it will be outdated very soon (and the "Ready for Migration" milestone - which this documentation targets - will be in a month).

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (tobru) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:153]: This might need some more explanation: Where do I create these objects?

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (tobru) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:595]: This part might need APUB-72 which is scheduled for the next sprint (aka next week).

**Review by tobru — CHANGES_REQUESTED** [review]
Did an initial review. The storage part is one which might need some more detailed explanation.

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (akosma) [reply docs/modules/ROOT/pages/tutorials/migration.adoc]: Removed the reference

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (akosma) [reply docs/modules/ROOT/pages/tutorials/migration.adoc]: Removed

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (akosma) [reply docs/modules/ROOT/pages/tutorials/migration.adoc]: Merged

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (akosma) [reply docs/modules/ROOT/pages/tutorials/migration.adoc]: Removed

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (akosma) [reply docs/modules/ROOT/pages/tutorials/migration.adoc]: Section removed

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (akosma) [reply docs/modules/ROOT/pages/tutorials/migration.adoc:567]: Should I add a reference to this, or just leave it like that, since it should be something available by the time we publish this?

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (akosma) [reply docs/modules/ROOT/pages/tutorials/migration.adoc]: Changed

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (tobru) [reply docs/modules/ROOT/pages/tutorials/migration.adoc:567]: No ref needed, just leave it as it is.

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (akosma) [reply docs/modules/ROOT/pages/tutorials/migration.adoc:125]: I'll wait for the review by @megian to rewrite this, as this section comes directly from the text of his ticket #36

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (ccremer) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:3]: :D
```suggestion
This document provides an overview of all the tasks required to move your applications from APPUiO Public to APPUiO Cloud.
```

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (ccremer) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:13]:
```suggestion
. Lowering your TTLs (Time To Live) 24 hours ahead of time will reduce the amount of time that your change takes to propagate. 
  You can limit downtime by lowering that value, so the caching servers will check back more frequently. 
  Typically, 60 seconds (1 minute) is a good idea.
```

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (akosma) [reply docs/modules/ROOT/pages/tutorials/migration.adoc:567]: Perfect! Resolving this conversation.

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (ccremer) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:21]:
```suggestion
In APPUiO Cloud only `Ingress` and `certificates.cert-manager.io/v1` objects can request certificates.
```

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (ccremer) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:55]: These annotations are wrong.
First, the ingress class on Cloud is not nginx, it would be HAproxy. But I believe it's not necessary if Haproxy is the default ingress class.
Secondly, we haven't activated the `kubernetes.io/tls-acme` annotation, only the `cert-manager.io/cluster-issuer`. 
```suggestion
metadata:
  annotations:
    cert-manager.io/cluster-issuer: letsencrypt-production <1>
  name: ${DEPLOYMENT_NAME}
  namespace: oauth-test
spec:
  rules:
  - host: ${ROUTE_HOST}
    http:
      paths:
      - backend:
          service:
            name: nginx
            port:
              number: 80
        path: /
        pathType: ImplementationSpecific
  tls:
  - hosts:
    - ${ROUTE_HOST}
    secretName: ${ROUTE_HOST}
--
<1> Annotation providing automatic provisioning of certificates on APPUiO Cloud.
```

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (ccremer) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:59]:
```suggestion
For reference, it was possible to configure TLS certificates in `Route` objects in APPUiO Public, using a manifest similar to the following. 
This is no longer valid in APPUiO Cloud.
```

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (ccremer) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:88]:
```suggestion
APPUiO Cloud has different requirements regarding storage.
This section provides more information.
```

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (ccremer) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:94]:
```suggestion
. The rollout strategy must change from `RollingUpdate` to `Recreate`, because mounting the volume at different nodes simultaneously isn't possible, or at least not with commonly used filesystems, such as https://en.wikipedia.org/wiki/Ext4[ext4] and https://en.wikipedia.org/wiki/XFS[XFS]. 
  https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#strategy[This page] in the Kubernetes documentation provides more information about this.
```

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (ccremer) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:96]:
```suggestion
. Add an affinity rule to run on the same node as the worker pod. 
  This allows to access the same RWO/block volume, because the filesystem is mounted on the node level, and allows a safe access to the same filesystem, because it's the same kernel.
```

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (ccremer) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:205]:
```suggestion
The second option for migrating your storage from APPUiO Public to APPUiO Cloud consists in using jobs.
The manifest below defines the objects required for this strategy.
```

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (ccremer) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:363]:
```suggestion
The third option to migrate your storage to APPUiO Cloud consists in using a Continuous Sync strategy. The main benefit of this approach is that files are replicated immediately after they're created.
If the destination pod dies, the sync pod also crashes, but is automatically restarted.
```

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (ccremer) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:529]:
```suggestion
Since APPUiO Cloud is based on OpenShift 4, there are new requirements for your container images.
This section contains all the required steps for adapting your images to the new environment.
```

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (ccremer) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:537]:
```suggestion
. Use the generated user token to authenticate to the registry on the command line.
  As the user token has enough privileges to read the image, a service account token isn't required.
```

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (ccremer) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:569]:
```suggestion
On OpenShift 4 it's also possible to find the token from https://oauth-openshift.apps.<cluster>/oauth/token/display and get read access; but this token doesn't grant enough privileges to write images.
  Therefore it's recommended to create a service account, and to grant access to `system:image-builders`, and finally to get the token from this service account.
```

**Review by ccremer — COMMENTED** [review]
Mostly styling suggestions re one-line-per sentence.
I did not look into the migration part itself (e.g. review the manifests)

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (megian) [reply docs/modules/ROOT/pages/tutorials/migration.adoc:125]: This part has to be installed on the destination side. Overall a pod must be running with the RWO volume mounted, the òc rsync` can connect to.

**Review by megian — APPROVED** [review]

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (corvus-ch) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:17]:
```suggestion
. There are different `CNAME` resource records to point to depending on which {zone} you migrat the application to; please refer to xref:zones.adoc[the {zone}s] reference document, where the corresponding `CNAME` entry is specified.
```

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (corvus-ch) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:100]: This is read as "an affinity". I am always uncertain if the `n` should be there in written text or not.

**Review by corvus-ch — CHANGES_REQUESTED** [review]
Oure style guide defines to have one sentence per line. Please check that rule is followed.

Hint: for "APPUiO Cloud" you can use `{product}` instead. It will be consistent and update should we change that for some reason.

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (ccremer) [reply docs/modules/ROOT/pages/tutorials/migration.adoc:100]: The general recommendation as I learned in university is to write `an` if the first letter is a vocal.
e.g. `an apple`, `an orange` but `a banana`

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (ccremer) [reply docs/modules/ROOT/pages/tutorials/migration.adoc:37]: The `# ` isn't necessary to annotate code

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (ccremer) [reply docs/modules/ROOT/pages/tutorials/migration.adoc:34]: This annotation doesn't seem to be actually required according to https://docs.appuio.cloud/user/how-to/getting-a-certificate.html

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (akosma) [reply docs/modules/ROOT/pages/tutorials/migration.adoc:13]: My eyes! My eyes!!

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (akosma) [reply docs/modules/ROOT/pages/tutorials/migration.adoc]: My eyes! My eyes!!

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (akosma) [reply docs/modules/ROOT/pages/tutorials/migration.adoc]: Merged

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (akosma) [reply docs/modules/ROOT/pages/tutorials/migration.adoc:59]: I do not understand what changed here.

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (akosma) [reply docs/modules/ROOT/pages/tutorials/migration.adoc:205]: Again, what's the purpose of these changes?

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (ccremer) [reply docs/modules/ROOT/pages/tutorials/migration.adoc:59]: As  @corvus-ch already explained: We follow the one-line-per-sentence style guide in asciidoc. So I just moved `This is no longer valid in APPUiO Cloud` to the next line.

**Comment by mhutter** [comment]
Note to self

- [ ] Can users verify whether they use outdated API versions?

**Comment by megian** [comment]
> Note to self
> 
>     * [ ]  Can users verify whether they use outdated API versions?

Potential tool https://github.com/doitintl/kube-no-trouble

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (akosma) [reply docs/modules/ROOT/pages/tutorials/migration.adoc:59]: We are not doing such a thing anywhere else; where is such a rule documented?

Comment (corvus-ch) [reply docs/modules/ROOT/pages/tutorials/migration.adoc:59]: ttps://syn.tools/syn/references/style-guide.html#_asciidoc_style. Yes, I understand that that the scope of that guide and this repository do not nessesarily overlap. Hower, we from Tarazed are used to follow it and it became the defacto standard also for all we do.

**Comment by corvus-ch** [comment]
> Can users verify whether they use outdated API versions?

And also potentially https://github.com/rikatz/kubepug

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (megian) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:104]:
```suggestion
To create an affinity rule for all the pods at once, you can use the https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/[app.kubernetes.io/component label].
```

File: docs/modules/ROOT/pages/tutorials/migration.adoc
Comment (megian) [inline docs/modules/ROOT/pages/tutorials/migration.adoc:129]:
```suggestion
The first option to migrate your storage from APPUiO Public to APPUiO Cloud consists in using `rsync`.
The manifests below create the required objects.
This part has to be installed on the destination side.
Overall a pod must be running with the RWO volume mounted, the `oc rsync` can connect to.
```

**Comment by akosma** [comment]
This PR is ready for merge, to be done Monday Nov. 22nd.

**Comment by akosma** [comment]
Merging

**Pull request merged**
